File: ebpf/ebpf_agent.c
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>
#include <linux/fs.h>
#include <linux/socket.h>
#include <linux/un.h>
#include <net/sock.h>

#define MIN(a,b) ({ __typeof__ (a) _a = (a); __typeof__ (b) _b = (b); _a < _b ? _a : _b; })

#define FILE_NAME_LEN 256
#define WRITE_BUF_MAX_LEN 2048
#define MOUNT_DATA_MAX_LEN 1024

// __ptrace_request 
#define PTRACE_ATTACH 16

// ===== СТРУКТУРЫ ДАННЫХ =====

// Для CVE-2022-0492 detection
struct write_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    int fd;
    char comm[TASK_COMM_LEN];
    char filename[FILE_NAME_LEN];
    char buf[WRITE_BUF_MAX_LEN];
    int len;
};

struct mount_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    char comm[TASK_COMM_LEN];
    char source[FILE_NAME_LEN];
    char target[FILE_NAME_LEN];
    char data[MOUNT_DATA_MAX_LEN];
};

struct unshare_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    char comm[TASK_COMM_LEN];
    int flags;
};

// Для Docker socket detection
struct connect_unix_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    int sockfd;
    char sock_filename[FILE_NAME_LEN];
    int retval;
};

// Для GDB attach detection
struct ptrace_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    u32 target_pid;
};

struct pwrite_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    int fd;
    char filename[FILE_NAME_LEN];
    unsigned char buf[WRITE_BUF_MAX_LEN];
    u64 len;
    u64 offset;
};

// Для load module detection
struct finit_module_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    char pathname[FILE_NAME_LEN];
};

// Для mknod detection
struct mknod_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    char comm[TASK_COMM_LEN];
    char pathname[FILE_NAME_LEN];
    int mode;
};

// Для proc/sys write detection
struct proc_sys_write_data_t {
    u32 pid;
    u32 ppid;
    u32 uid;
    int fd;
    char filename[FILE_NAME_LEN];
    unsigned char buf[WRITE_BUF_MAX_LEN];
    u64 len;
};

struct opened_file_data_t {
    int fd;
    char pathname[FILE_NAME_LEN];
};

// ===== PERF OUTPUT БУФЕРЫ =====
// таблицы для передачи событий из ядра в пользовательское пространство
// https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#2-bpf_perf_output
BPF_PERF_OUTPUT(write_events);
BPF_PERF_OUTPUT(mount_events);
BPF_PERF_OUTPUT(unshare_events);
BPF_PERF_OUTPUT(connect_unix_events);
BPF_PERF_OUTPUT(ptrace_events);
BPF_PERF_OUTPUT(pwrite_events);
BPF_PERF_OUTPUT(finit_module_events);
BPF_PERF_OUTPUT(mknod_events);
BPF_PERF_OUTPUT(proc_sys_write_events);

// ===== МАССИВЫ ДЛЯ ОБХОДА ОГРАНИЧЕНИЙ СТЕКА =====
// массивы для обхода ограничения bpf в 512 байт на стэк
// одно из обсуждений с этой "проблемой"
// https://github.com/iovisor/bcc/issues/2306
BPF_ARRAY(write_data, struct write_data_t, 1);
BPF_ARRAY(mount_data, struct mount_data_t, 1);
BPF_ARRAY(connect_data, struct connect_unix_data_t, 1);
BPF_ARRAY(pwrite_data, struct pwrite_data_t, 1);
BPF_ARRAY(proc_sys_write_data, struct proc_sys_write_data_t, 1);

// ===== ХЭШ-ТАБЛИЦЫ =====
// промежуточная таблица для сохранения имени файла между входом и выходом из openat
BPF_HASH(temp_opened_files, u64, const char *);
// промежуточная таблица для сохранения имени файла между входом и выходом из dup
BPF_HASH(temp_dup_files, u64, int);
// итоговая таблица сопастовления fd и имени файла
BPF_HASH(opened_files, u64, struct opened_file_data_t);

// ===== ФУНКЦИИ ОТСЛЕЖИВАНИЯ ФАЙЛОВ =====
// отслеживаем вызовы openat для "резолва" имен файлов в последующих вызовах ( в частности во write )
// https://man7.org/linux/man-pages/man2/openat.2.html
int syscall__openat(struct pt_regs *ctx, int dirfd, const char *pathname, int flags, mode_t mode)
{
    u64 file_id = bpf_get_current_pid_tgid();

    // используем id процесса и потока в качестве "идентификатора" файла находящегося в процессе открытия
    // и сохраняем имя файла для последующего мапинга с полученным при выходе дескриптором
    // далее связка "имя файла" -> номер дескриптора будет использована в других вызовах ( напр. write)
    // для получения имени файла, к которому осуществляется обращение

    temp_opened_files.update(&file_id, &pathname);
    return 0;
}

int syscall__openat_ret(struct pt_regs *ctx)
{
    // используем id процесса и потока в качестве "идентификатора" файла находящегося в процессе открытия
    u64 temp_file_id = bpf_get_current_pid_tgid();

    // получаем возвращаемое значение
    // -1 при ошибке
    // иначе - вернется номер открытого дескриптора
    int retval = PT_REGS_RC(ctx);
    
    // ищем ранее сохраненное имя файла с полученным выше "идентификатором" temp_file_id
    const char **pathname = temp_opened_files.lookup(&temp_file_id);
    if (!pathname)
        return 0;

    // при ошибке более не отслеживаем этот файл и удаляем из хэш-таблицы
    if ( retval == -1 ) {
        temp_opened_files.delete(&temp_file_id);
        return 0;
    }

    // получаем "итоговый" id файла из id потока и номера полученного дескриптора
    u64 file_id = (bpf_get_current_pid_tgid() & 0xFFFFFFFF00000000) + retval;

    struct opened_file_data_t opened_file = {};
    
    opened_file.fd = retval;
    bpf_probe_read_user(opened_file.pathname, sizeof(opened_file.pathname), *pathname);

    // сохраняем в хэш-таблицу хранящую имена файлов и дескрипторов открытых файлов
    opened_files.update(&file_id, &opened_file);

    // удаляем из хэш-таблицы хранящую файлы в процессе открытия
    temp_opened_files.delete(&temp_file_id);

    return 0;
}

// https://man7.org/linux/man-pages/man2/dup.2.html
int syscall__dup2(struct pt_regs *ctx, int oldfd, int newfd) {
    // используем id процесса и потока в качестве "идентификатора" файла находящегося в процессе дублирования
    // и сохраняем номер переданного ( старого ) дескриптора для последующего мапинга с полученным при выходе дублированным дескриптором
    // далее связка "номер старого дескриптора" -> "номер нового дескриптора" будет использована при выходе из вызова
    // для создания записи в хэш-таблицы открытых файлов с имени файла для старого дескриптора, но с новым номером дескриптора ( дублированным )
    
    u64 temp_old_id = bpf_get_current_pid_tgid();
    temp_dup_files.update(&temp_old_id, &oldfd);
    return 0;
}

int syscall__dup2_ret(struct pt_regs *ctx) {
    // получаем возвращаемое значение
    // -1 при ошибке
    // иначе - номер дублированного дискриптора
    int retval = PT_REGS_RC(ctx);

    if ( retval == -1 ) {
        return 0;
    }

    // используем id процесса и потока в качестве "идентификатора" файла находящегося в процессе дублирования
    u64 temp_dup_file_id = bpf_get_current_pid_tgid();

    // получаем номер дублируемого дескриптора
    int *old_file_id = temp_dup_files.lookup(&temp_dup_file_id);

    // если дублируемый дескриптор найден, то создаем запись в хэш-таблице открытых файлов
    if (old_file_id) {
        u64 pid_tgid = bpf_get_current_pid_tgid();

        // получаем id для записи хранящей имя файла ассоциированного с дублируемым дескриптором в хэш-таблице открытых файлов
        u64 old_opened_file_id = (pid_tgid & 0xFFFFFFFF00000000) + *old_file_id;

        // получаем запись из хэш-таблицы открытых файлов
        struct opened_file_data_t *target_file = opened_files.lookup(&old_opened_file_id);
        if (!target_file)
            return 0;
        
        // получаем "итоговый" id файла из id потока и номера полученного дескриптора
        u64 dup_file_id = (pid_tgid & 0xFFFFFFFF00000000) + retval;
        
        struct opened_file_data_t dup_file = {};

        // сохраняем новый дискриптор
        dup_file.fd = retval;

        // и старое имя файла
        bpf_probe_read_kernel(dup_file.pathname, sizeof(dup_file.pathname), target_file->pathname);

        // сохраняем в хэш-таблицу открытых файлов
        opened_files.update(&dup_file_id, &dup_file);

        // удаляем из хэш-таблицы хранящую дескрипторы в процессе дублирования
        temp_dup_files.delete(&temp_dup_file_id);
    }

    return 0;
}

// https://man7.org/linux/man-pages/man2/close.2.html
int syscall__close(struct pt_regs *ctx, int fd)
{
    u64 file_id = ( bpf_get_current_pid_tgid() & 0xFFFFFFFF00000000 ) + fd;

    struct opened_file_data_t *opened_file = opened_files.lookup(&file_id);
    if (opened_file)
        opened_files.delete(&file_id);

    return 0;
}

// ===== CVE-2022-0492 DETECTION =====
// https://man7.org/linux/man-pages/man2/write.2.html
int syscall__write(struct pt_regs *ctx, int fd, void *buf, u64 buf_len)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u64 file_id  = ( pid_tgid & 0xFFFFFFFF00000000 ) + fd;

    struct opened_file_data_t *opened_file = opened_files.lookup(&file_id);
    if (!opened_file) return 0;

    int index = 0;

    struct write_data_t *data = write_data.lookup(&index);
    if (data == NULL) return 1;
    
    memset(data->filename, 0, FILE_NAME_LEN);

    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    data->pid  = pid_tgid >> 32;
    data->ppid = task->real_parent->tgid;
    data->fd   = fd;
    data->uid  = bpf_get_current_uid_gid() & 0xffffffff;
    u32 result_len = MIN(WRITE_BUF_MAX_LEN, buf_len);
    bpf_probe_read_user(data->buf, result_len, buf);
    data->len = result_len;
    bpf_get_current_comm(data->comm, sizeof(data->comm));

    bpf_probe_read_kernel(data->filename, sizeof(data->filename), opened_file->pathname);

    write_events.perf_submit(ctx, data, sizeof(*data));

    return 0;
}

// https://man7.org/linux/man-pages/man2/mount.2.html
int syscall__mount(struct pt_regs *ctx,
    const char *source, const char *target,
    const char *filesystemtype, unsigned long mountflags,
    const void *m_data)
{
    int index = 0;

    struct mount_data_t *data = mount_data.lookup(&index);
    if (data == NULL) return 1;
    
    memset(data->source, 0, FILE_NAME_LEN);
    memset(data->target, 0, FILE_NAME_LEN);

    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    data->pid  = bpf_get_current_pid_tgid() >> 32;
    data->ppid = task->real_parent->tgid;
    data->uid  = bpf_get_current_uid_gid() & 0xFFFFFFFF;
    bpf_get_current_comm(data->comm, sizeof(data->comm));
    bpf_probe_read_user(data->source, sizeof(data->source), source);
    bpf_probe_read_user(data->target, sizeof(data->target), target);
    bpf_probe_read_user_str(data->data, sizeof(data->data), m_data);

    mount_events.perf_submit(ctx, data, sizeof(*data));

    return 0;
}

int syscall__unshare(struct pt_regs *ctx, int flags){
    struct unshare_data_t unshare_data = {};

    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    unshare_data.pid  = bpf_get_current_pid_tgid() >> 32;
    unshare_data.ppid = task->real_parent->tgid;
    unshare_data.uid  = bpf_get_current_uid_gid() & 0xFFFFFFFF;
    unshare_data.flags = flags;
    bpf_get_current_comm(&unshare_data.comm, sizeof(unshare_data.comm));

    unshare_events.perf_submit(ctx, &unshare_data, sizeof(unshare_data));

    return 0;
}

// ===== DOCKER SOCKET DETECTION =====
// https://man7.org/linux/man-pages/man2/connect.2.html
int syscall__connect(struct pt_regs *ctx, int sockfd, struct sockaddr *saddr, u64 addrlen)
{
    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    u64 pid_tgid  = bpf_get_current_pid_tgid();
    u32 pid       = pid_tgid >> 32;
    u32 ppid      = task->real_parent->tgid;
    u32 uid       = bpf_get_current_uid_gid() & 0xFFFFFFFF;

    if ( saddr->sa_family == 1 ) // sa_family == UNIX
    {
        struct sockaddr_un *saddr_un = (struct sockaddr_un*) saddr;

        if ( saddr_un->sun_path[0] == 0) // sun_path пустой
            return 0;
        
        int index = 0;

        struct connect_unix_data_t *data = connect_data.lookup(&index);
        if (data == NULL) return 1;
        
        data->pid     = pid;
        data->ppid    = ppid;
        data->uid     = uid;
        data->sockfd  = sockfd;

        bpf_probe_read_user(data->sock_filename, sizeof(data->sock_filename), saddr_un->sun_path);

        u64 socket_id = ( pid_tgid & 0xFFFFFFFF00000000 ) + sockfd;

        struct opened_file_data_t opened_socket = {};

        opened_socket.fd = sockfd;
        bpf_probe_read_user(opened_socket.pathname, sizeof(opened_socket.pathname), saddr_un->sun_path);

        opened_files.update(&socket_id, &opened_socket);

        connect_unix_events.perf_submit(ctx, data, sizeof(*data));
    }

    return 0;
}

// https://man7.org/linux/man-pages/man2/sendto.2.html
int syscall__sendto(struct pt_regs *ctx, int fd, void *buf, u64 buf_len, int flags, struct sockaddr *dest_addr, int addrlen)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u64 file_id  = ( pid_tgid & 0xFFFFFFFF00000000 ) + fd;

    struct opened_file_data_t *opened_file = opened_files.lookup(&file_id);
    if (!opened_file) return 0;

    int index = 0;

    struct write_data_t *data = write_data.lookup(&index);
    if (data == NULL) return 1;
    
    memset(data->filename, 0, FILE_NAME_LEN);

    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    data->pid  = pid_tgid >> 32;
    data->ppid = task->real_parent->tgid;
    data->fd   = fd;
    data->uid  = bpf_get_current_uid_gid() & 0xffffffff;
    u32 result_len = MIN(WRITE_BUF_MAX_LEN, buf_len);
    bpf_probe_read_user(data->buf, result_len, buf);
    data->len = result_len;
    bpf_get_current_comm(data->comm, sizeof(data->comm));

    bpf_probe_read_kernel(data->filename, sizeof(data->filename), opened_file->pathname);

    write_events.perf_submit(ctx, data, sizeof(*data));

    return 0;
}

// ===== GDB ATTACH DETECTION =====
// https://man7.org/linux/man-pages/man2/ptrace.2.html
int syscall__ptrace(struct pt_regs *ctx,
    u32 request, u32 pid,
    void *addr, void *data)
{
    // отслеживаем только события PTRACE_ATTACH
    if ( request != PTRACE_ATTACH )
        return 0;

    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    u64 pid_tgid = bpf_get_current_pid_tgid();

    struct ptrace_data_t ptrace_data = {};

    ptrace_data.pid  = pid_tgid >> 32;
    ptrace_data.ppid = task->real_parent->tgid;
    ptrace_data.uid  = bpf_get_current_uid_gid() & 0xffffffff;

    ptrace_data.target_pid = pid;

    ptrace_events.perf_submit(ctx, &ptrace_data, sizeof(ptrace_data));

    return 0;
}

// https://man7.org/linux/man-pages/man2/pwrite64.2.html
int syscall__pwrite64(struct pt_regs *ctx, int fd, void *buf, u64 buf_len, u64 offset)
{   
    int index = 0;
    struct pwrite_data_t *data = pwrite_data.lookup(&index);

    if (data == NULL) return 1;

    memset(data->filename, 0, FILE_NAME_LEN);

    u64 pid_tgid  = bpf_get_current_pid_tgid();

    // формируем id для хэш-таблице хранящей "маппинг" дескрипторов и имен файлов
    u64 file_id = ( pid_tgid & 0xFFFFFFFF00000000 ) + fd;

    struct opened_file_data_t *opened_file = opened_files.lookup(&file_id);
    if (!opened_file)
        return 0;

    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    data->pid  = pid_tgid >> 32;
    data->ppid = task->real_parent->tgid;
    data->fd   = fd;
    data->uid  = bpf_get_current_uid_gid() & 0xffffffff;
    bpf_probe_read_kernel(data->filename, sizeof(data->filename), opened_file->pathname);
    u32 result_len = MIN(WRITE_BUF_MAX_LEN, buf_len);
    bpf_probe_read_user(data->buf, result_len, buf);
    data->len    = result_len;
    data->offset = offset;

    // выводим данные в userspace
    pwrite_events.perf_submit(ctx, data, sizeof(struct pwrite_data_t));

    return 0;
}

// ===== LOAD MODULE DETECTION =====
// https://man7.org/linux/man-pages/man2/finit_module.2.html
int syscall__finit_module(struct pt_regs *ctx, int fd, const char *param_values, int flags){
    struct finit_module_data_t finit_module_data = {};

    u64 pid_tgid  = bpf_get_current_pid_tgid();
    u64 file_id   = ( pid_tgid & 0xFFFFFFFF00000000 ) + fd;

    struct opened_file_data_t *opened_file = opened_files.lookup(&file_id);
    if (!opened_file)
        return 0;
    
    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    finit_module_data.pid  = pid_tgid >> 32;
    finit_module_data.ppid = task->real_parent->tgid;
    finit_module_data.uid  = bpf_get_current_uid_gid() & 0xffffffff;

    bpf_probe_read_kernel(finit_module_data.pathname, sizeof(finit_module_data.pathname), opened_file->pathname);

    finit_module_events.perf_submit(ctx, &finit_module_data, sizeof(finit_module_data));

    return 0;
}

// ===== MKNOD DETECTION =====
// https://man7.org/linux/man-pages/man2/mknod.2.html
int syscall__mknod(struct pt_regs *ctx, int dirfd, const char *pathname, mode_t mode, dev_t dev)
{
    struct mknod_data_t mknod_data = {};

    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    mknod_data.pid  = bpf_get_current_pid_tgid() >> 32;
    mknod_data.ppid = task->real_parent->tgid;
    mknod_data.uid  = bpf_get_current_uid_gid() & 0xFFFFFFFF;
    bpf_get_current_comm(&mknod_data.comm, sizeof(mknod_data.comm));
    bpf_probe_read_user(mknod_data.pathname, sizeof(mknod_data.pathname), pathname);
    mknod_data.mode = mode;

    mknod_events.perf_submit(ctx, &mknod_data, sizeof(mknod_data));

    return 0;
}

// ===== PROC/SYS WRITE DETECTION =====

int syscall__proc_sys_write(struct pt_regs *ctx, int fd, void *buf, u64 buf_len)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u64 file_id  = ( pid_tgid & 0xFFFFFFFF00000000 ) + fd;

    struct opened_file_data_t *opened_file = opened_files.lookup(&file_id);
    if (!opened_file) return 0;

    int index = 0;
    struct proc_sys_write_data_t *data = proc_sys_write_data.lookup(&index);
    
    if (data == NULL) return 1;
    
    memset(data->filename, 0, FILE_NAME_LEN);

    struct task_struct *task;
    task = (struct task_struct *)bpf_get_current_task();

    data->pid  = pid_tgid >> 32;
    data->ppid = task->real_parent->tgid;
    data->fd = fd;
    data->uid  = bpf_get_current_uid_gid() & 0xFFFFFFFF;
    u32 result_len = MIN(WRITE_BUF_MAX_LEN, buf_len);
    bpf_probe_read_user(data->buf, result_len, buf);
    data->len = result_len;
    bpf_probe_read_kernel(data->filename, sizeof(data->filename), opened_file->pathname);

    proc_sys_write_events.perf_submit(ctx, data, sizeof(*data));

    return 0;
} 

File: ebpf/ebpf_agent.py
#!/usr/bin/env python3

from bcc import BPF
from util import *
import stat
import os

# Имя файла с кодом kernel-части
current_dir = os.path.dirname(os.path.abspath(__file__))
BPF_FILENAME = os.path.join(current_dir, "ebpf_agent.c")

try:
    with open(BPF_FILENAME, "r") as bpf_file:
        bpf_text = bpf_file.read()
except:
    print(f"ERROR: cant open {BPF_FILENAME}")
    exit(1)

# Инициализируем BPF
bpf = BPF(text=bpf_text, cflags=["-Wno-macro-redefined"])

# Получаем имена для системных вызовов
openat_fname        = bpf.get_syscall_fnname("openat")
dup2_fname          = bpf.get_syscall_fnname("dup2")
close_fnname        = bpf.get_syscall_fnname("close")
write_fnname        = bpf.get_syscall_fnname("write")
mount_fnname        = bpf.get_syscall_fnname("mount")
unshare_fnname      = bpf.get_syscall_fnname("unshare")
connect_fnname      = bpf.get_syscall_fnname("connect")
sendto_fnname       = bpf.get_syscall_fnname("sendto")
ptrace_fname        = bpf.get_syscall_fnname("ptrace")
pwrite_fname        = bpf.get_syscall_fnname("pwrite64")
finit_module_fnname = bpf.get_syscall_fnname("finit_module")
mknod_fnname        = bpf.get_syscall_fnname("mknodat")

# Прикрепляемся к системным вызовам
bpf.attach_kprobe(event=openat_fname,        fn_name="syscall__openat")
bpf.attach_kretprobe(event=openat_fname,     fn_name="syscall__openat_ret")
bpf.attach_kprobe(event=dup2_fname,          fn_name="syscall__dup2")
bpf.attach_kretprobe(event=dup2_fname,       fn_name="syscall__dup2_ret")
bpf.attach_kprobe(event=close_fnname,        fn_name="syscall__close")

# CVE-2022-0492 detection
bpf.attach_kprobe(event=write_fnname,        fn_name="syscall__write")
bpf.attach_kprobe(event=mount_fnname,        fn_name="syscall__mount")
bpf.attach_kprobe(event=unshare_fnname,      fn_name="syscall__unshare")

# Docker socket detection
bpf.attach_kprobe(event=connect_fnname,      fn_name="syscall__connect")
bpf.attach_kprobe(event=sendto_fnname,       fn_name="syscall__sendto")

# GDB attach detection
bpf.attach_kprobe(event=ptrace_fname,        fn_name="syscall__ptrace")
bpf.attach_kprobe(event=pwrite_fname,        fn_name="syscall__pwrite64")

# Load module detection
bpf.attach_kprobe(event=finit_module_fnname, fn_name="syscall__finit_module")

# Mknod detection
bpf.attach_kprobe(event=mknod_fnname,        fn_name="syscall__mknod")

# Proc/sys write detection - используем обычный write с фильтрацией
bpf.attach_kprobe(event=write_fnname,        fn_name="syscall__proc_sys_write")

# Глобальные переменные для отслеживания состояния
ptraced_pids = {}

# ===== ОБРАБОТЧИКИ СОБЫТИЙ =====

# CVE-2022-0492 Detection handlers
def handle_write_event(cpu, data, size):
    event = bpf["write_events"].event(data)

    proc = Process(event)

    if not proc.in_container():
        return
    
    target = event.filename.decode("ascii", errors="ignore")

    if target.endswith("/release_agent") or target.endswith("/notify_on_release"):
        write_buf = event.buf[:event.len].decode("ascii", errors='ignore')
        message = f"[CVE-2022-0492] writing to {target}\n[data]\n{write_buf}"
        print_event_message(proc, message)

def handle_mount_event(cpu, data, size):
    event = bpf["mount_events"].event(data)
    
    proc = Process(event)

    if not proc.from_container():
        return
    
    source = event.source.decode("ascii")
    target = event.target.decode("ascii")
    data_str = event.data.decode("ascii", errors="ignore")
    
    if source == "cgroup" and "rdma" in data_str:
        message = f"[CVE-2022-0492] mounting {source} to {target} [options/data]: {data_str}"
        print_event_message(proc, message)
    else:
        # Общее отслеживание mount операций
        if not data_str:
            data_str = "none"
        message = f"[MOUNT] mounting {source} to {target} [options/data]: {data_str}"
        print_event_message(proc, message)

CLONE_NEWCGROUP = 0x02000000 # new cgroup namespace
CLONE_NEWUSER   = 0x10000000 # new user namespace

def handle_unshare_event(cpu, data, size):
    event = bpf["unshare_events"].event(data)

    proc = Process(event)
    if not proc.from_container():
        return
    
    flags = event.flags

    if (flags & CLONE_NEWUSER) and (flags & CLONE_NEWCGROUP):
        message = f"[CVE-2022-0492] unshare ({flags}) with user and cgroup namespace from container"
        print_event_message(proc, message)

# Docker Socket Detection handlers
def handle_connect_unix_event(cpu, data, size):
    event = bpf["connect_unix_events"].event(data)

    proc = Process(event)

    sock_name = event.sock_filename.decode("ascii")

    if sock_name.endswith("/docker.sock") and proc.in_container():
        message = f"[DOCKER-SOCK] connecting to {sock_name}"
        print_event_message(proc, message)

# GDB Attach Detection handlers
def handle_ptrace_event(cpu, data, size):
    global ptraced_pids

    event = bpf["ptrace_events"].event(data)

    proc = Process(event)

    target_pid = event.target_pid
    target_exe = get_pid_realpath(target_pid)

    message = f"[GDB-ATTACH] ptrace attach to pid: {target_pid} ( {target_exe} )"

    ptraced_pids[target_pid] = target_exe

    print_event_message(proc, message)
    
def handle_pwrite_event(cpu, data, size):
    event = bpf["pwrite_events"].event(data)

    proc = Process(event)

    target = event.filename.decode("ascii", errors='ignore')

    if not target.endswith("/mem"):
        return
    
    target_pid = int(target.split("/")[-2])
    ptraced_pid = ptraced_pids.get(target_pid)

    if not ptraced_pid:
        return 
    
    pwrite_buf = bytes(event.buf)[:event.len].hex()
    
    message = f"[GDB-ATTACH] pwrite to {target} [ {pwrite_buf} ]"

    print_event_message(proc, message)

# Load Module Detection handler
def handle_finit_module(cpu, data, size):
    event = bpf["finit_module_events"].event(data)

    proc = Process(event)

    if not proc.from_container():
        return
    
    pathname = event.pathname.decode("ascii")
    
    message = f"[LOAD-MODULE] loading {pathname} module from container"

    print_event_message(proc, message)

# Mknod Detection handler
def handle_mknod_event(cpu, data, size):
    event = bpf["mknod_events"].event(data)

    proc = Process(event)

    pathname = event.pathname.decode("ascii")
    mode = event.mode

    if stat.S_ISBLK(mode) and proc.from_container():
        message = f"[MKNOD] creating block device {pathname} from container"
        print_event_message(proc, message)

# Proc/Sys Write Detection handler
def handle_proc_sys_write_event(cpu, data, size):
    event = bpf["proc_sys_write_events"].event(data)

    proc = Process(event)

    target = event.filename.decode("ascii")

    if target.endswith("/core_pattern") or target.endswith("/uevent_helper"):
        write_buf = bytes(event.buf)[:event.len].decode("ascii", errors='ignore')
        message = f"[PROC-SYS-WRITE] writing to {target}\n[data]\n{write_buf}"
        print_event_message(proc, message)

# Дополнительная обработка для Docker socket write events
def handle_docker_write_event(cpu, data, size):
    event = bpf["write_events"].event(data)

    proc = Process(event)

    target = event.filename.decode("ascii", errors='ignore')

    if target.endswith("/docker.sock") and proc.in_container():
        write_buf = event.buf[:event.len].decode("ascii", errors='ignore')

        if '/containers/create' in write_buf:
            if '"Privileged":true' in write_buf:
                message = f"[DOCKER-SOCK] creating privileged container from container\n[data]\n{write_buf}"
                print_event_message(proc, message)
            elif '"Binds":[' in write_buf:
                message = f"[DOCKER-SOCK] creating container with \"binds\" from container\n[data]\n{write_buf}"
                print_event_message(proc, message)

# Объединенный обработчик для write событий
def handle_combined_write_event(cpu, data, size):
    handle_write_event(cpu, data, size)
    handle_docker_write_event(cpu, data, size)

# Прикрепляем обработчики к событиям
bpf["write_events"].open_perf_buffer(handle_combined_write_event)
bpf["mount_events"].open_perf_buffer(handle_mount_event)
bpf["unshare_events"].open_perf_buffer(handle_unshare_event)
bpf["connect_unix_events"].open_perf_buffer(handle_connect_unix_event)
bpf["ptrace_events"].open_perf_buffer(handle_ptrace_event)
bpf["pwrite_events"].open_perf_buffer(handle_pwrite_event)
bpf["finit_module_events"].open_perf_buffer(handle_finit_module)
bpf["mknod_events"].open_perf_buffer(handle_mknod_event)
bpf["proc_sys_write_events"].open_perf_buffer(handle_proc_sys_write_event)

print("=== COMBINED eBPF SECURITY MONITOR ===")
print("Monitoring:")
print("  - CVE-2022-0492 (cgroup release_agent escape)")
print("  - Docker socket access from containers")
print("  - Mount operations")
print("  - GDB attach and memory write operations")
print("  - Kernel module loading")
print("  - Block device creation (mknod)")
print("  - Proc/sys write operations")
print("")

print("%-12s %-7s %-14s %-7s %-40s      %-16s %s" % 
("CONTAINER_ID", "PID", "PCONTAINER_ID", "PPID", "PCOMM", "COMM", "MESSAGE"))

# В бесконечном цикле получаем и обрабатываем поступающие события
while True:
    try:
        bpf.perf_buffer_poll()
    except KeyboardInterrupt:
        print("\nShutting down...")
        exit() 